src/lsdb/core/crossmatch/kdtree_match.py

```diff
@@ -139,3 +153,65 @@ def _lon_lat_to_xyz(lon: npt.NDArray[np.float64], lat: npt.NDArray[np.float64])
     z = np.sin(lat)
 
     return np.stack([x, y, z], axis=1)
+
+
+def _get_chord_distance(radius_arcsec: float) -> float:
+    """Calculates the distance between two points on the surface of the unit sphere,
+    for a given radius, in arcseconds
+
+    Args:
+        radius_arcsec (float): Tbe radius, in arcseconds
+
+    Returns:
+        The chord distance between the two points on the unit sphere.
+    """
+    radius_degrees = radius_arcsec / 3600.0
+    return 2.0 * math.sin(math.radians(0.5 * radius_degrees))
+
+
+def _query_min_max_neighbors(
+    tree: KDTree,
+    left_xyz: npt.NDArray[np.float64],
+    right_xyz: npt.NDArray[np.float64],
+    n_neighbors: int,
+    min_distance: float,
+    max_distance: float,
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Finds `n_neighbors` within a distance range for all points in a pair of partitions"""
+    left_tree = KDTree(left_xyz, compact_nodes=True, balanced_tree=True, copy_data=False)
+
+    # Find the number of neighbors within the minimum distance threshold
+    len_too_close_neighbors = np.zeros(left_xyz.shape[0], dtype=np.int64)
+    left_indices_too_close = left_tree.sparse_distance_matrix(
+        tree, max_distance=min_distance, output_type="ndarray"
+    )["i"]
+    unique, counts = np.unique(left_indices_too_close, return_counts=True)
+    len_too_close_neighbors[unique] = counts
+
+    # Make sure we don't ask for more neighbors than there are points
+    n_neighbors_to_request = min(n_neighbors + max(len_too_close_neighbors), len(right_xyz))
```

Comment:
I think the mask logic below fails if there are fewer points than we want to ask for

Reply:
That's right! I pushed a fix and added a unit test to make sure the mask is successfully generated for this case.